src/sdk.hpp

```diff
@@ -0,0 +1,187 @@
+#define FE_n13(what, arg, a, ...)  what(12, arg, a) FE_CALLITn13(FE_n12,(what, arg, ##__VA_ARGS__))
+#define FE_n14(what, arg, a, ...)  what(13, arg, a) FE_CALLITn14(FE_n13,(what, arg, ##__VA_ARGS__))
+#define FE_n15(what, arg, a, ...)  what(14, arg, a) FE_CALLITn15(FE_n14,(what, arg, ##__VA_ARGS__))
+#define FE_n16(what, arg, a, ...)  what(15, arg, a) FE_CALLITn16(FE_n15,(what, arg, ##__VA_ARGS__))
+#define FE_n17(what, arg, a, ...)  what(16, arg, a) FE_CALLITn17(FE_n16,(what, arg, ##__VA_ARGS__))
+#define FE_n18(what, arg, a, ...)  what(17, arg, a) FE_CALLITn18(FE_n17,(what, arg, ##__VA_ARGS__))
+#define FE_n19(what, arg, a, ...)  what(18, arg, a) FE_CALLITn19(FE_n18,(what, arg, ##__VA_ARGS__))
+#define FE_n20(what, arg, a, ...)  what(19, arg, a) FE_CALLITn20(FE_n19,(what, arg, ##__VA_ARGS__))
+#define FE_n21(what, arg, a, ...)  what(20, arg, a) FE_CALLITn21(FE_n20,(what, arg, ##__VA_ARGS__))
+#define FE_n22(...)           ERROR: FOR_EACH only supports up to 21 arguments
+#define FE_GET_MACRO(_0,_1,_2,_3,_4,_5,_6,_7,_8,_9,_10,_11,_12,_13,_14,_15,_16,_17,_18,_19,_20,_21,_22,NAME,...) NAME
+#define FOR_EACH(what, arg, ...) FE_CALLITn01(FE_GET_MACRO(_0, ##__VA_ARGS__,FE_n22,FE_n21,FE_n20,FE_n19, \
+                            FE_n18,FE_n17,FE_n16,FE_n15,FE_n14,FE_n13,FE_n12,FE_n11,FE_n10,FE_n09,\
+FE_n08,FE_n07,FE_n06,FE_n05,FE_n04,FE_n03,FE_n02,FE_n01,FE_n00), (what, arg, ##__VA_ARGS__))
+
+#define DECLTYPE(idx, _, property) \
+IF_ELSE(idx != 0)(decltype(property),)(decltype(property))
+
+#define CASE(idx, arg, property) \
+    case change::property: {\
+        auto p = property_change_##property(); \
+        realm::sdk::set_change_values<property_change_##property, decltype(property)>(p, std::get<0>(arg), std::get<1>(arg)); \
+        return p; \
+    }
+
+#define IF_NAME_EQUAL_TO_PROPERTY_RETURN_CHANGE(idx, name, property) \
+if (name == #property) { \
+    return Property::property; \
+}
+
+/**
+ MARK: REALM_PRIMARY_KEY
+ Wrap a cpp type in a Realm Property specialization with a trait marking that it is the primary key.
+ */
+#define REALM_PRIMARY_KEY(arg)\
+    using primary_key_type = arg; \
+    realm::sdk::Property<arg, realm::sdk::util::type_traits::primary_key>
+
+/**
+ MARK: REALM
+ Wrap a cpp type in a Realm Property specialization.
+ For EmbeddedObjects, this wrapping will yield the original type, so that
+ users can access its fields directly.
+ */
+
+#define REALM(arg) \
+    realm::sdk::util::conditional_property_type<arg>::type
+
+/**
+ MARK: REALM_EXPORT
+ Add metadata to an Object about the exported properties. This will add:
+ - a method that provides access to a tuple of the exported properties
+ - an inlined Schema generated based on the exported properties
+ - an enum listing the Properties so that user's can better understand what has changed (during observation)
+ - an observe method to allow Object observation
+ */
+#define REALM_EXPORT(...) \
```

Comment:
Why this macro? It seems that all of the stuff it does can be achieved with a `static constexpr` function defined by the user.